services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/actions/PrepareFlowAction.java

```diff
@@ -66,28 +66,18 @@ public ValidateFlowAction(PersistenceManager persistenceManager, FlowOperationsD
                 new HashSet<>(Optional.ofNullable(context.getPathsToReroute()).orElse(emptySet()));
         dashboardLogger.onFlowPathReroute(flowId, pathsToReroute, context.isForceReroute());
 
+        // alter all affected flow's fields here
         Flow flow = persistenceManager.getTransactionManager().doInTransaction(() -> {
-            Flow foundFlow = getFlow(flowId, FetchStrategy.NO_RELATIONS);
-            if (foundFlow.getStatus() == FlowStatus.IN_PROGRESS) {
-                throw new FlowProcessingException(ErrorType.REQUEST_INVALID,
-                        format("Flow %s is in progress now", flowId));
-            }
-
-            stateMachine.setOriginalFlowStatus(foundFlow.getStatus());
-            stateMachine.setOriginalEncapsulationType(foundFlow.getEncapsulationType());
-            stateMachine.setRecreateIfSamePath(!foundFlow.isActive() || context.isForceReroute());
-
-            flowRepository.updateStatus(foundFlow.getFlowId(), FlowStatus.IN_PROGRESS);
-            return foundFlow;
+            Flow targetFlow = getFlow(flowId);
+
+            targetFlow.setEncapsulationType(determineFlowEncapsulation(targetFlow));
+
+            flowRepository.createOrUpdate(targetFlow);
+
+            return targetFlow;
         });
 
-        featureTogglesRepository.find().ifPresent(featureToggles ->
-                Optional.ofNullable(featureToggles.getFlowsRerouteUsingDefaultEncapType()).ifPresent(toggle -> {
-                    if (toggle) {
-                        stateMachine.setNewEncapsulationType(
-                                kildaConfigurationRepository.get().getFlowEncapsulationType());
-                    }
-                }));
+        stateMachine.setRecreateIfSamePath(! stateMachine.getOriginalFlow().isActive() || context.isForceReroute());
```

Comment:
Will the flow be in a consistent state if the action updates the encapsulation type first and then finds the same path and skips the reroute process?